forc-plugins/forc-node/src/chain_config.rs

```diff
@@ -0,0 +1,76 @@
+//! This module creates copies of chain configs on the disk so that `forc-node`
+//! will always have a pinned instance of a chain config for given `Mode`.
```

Comment:
Do we actually want to do this? It makes sense for local, but I'm not sure user copies are a good idea for testnet & mainnet.

If/when we need to update the network config (gas cost, change privileged addresses, etc), we will publish an update to the `chain_config.json` and/or the `state_config.json` files. All nodes must have matching values in those files in order to connect to the network.

I can't think of a use-case for users having different values for those files on deployed networks, and I think we could cause confusion during network upgrades if we allow users to have local copies of them ("why won't my node connect?!?"). Though for running the "local" network, local copies make a ton of sense IMO.

Reply:
I realize the wording of the comment misses an aspect of what happens here so first a little clarification:

This module basically writes the embedded (I embed the chain config folders into the forc-node binary) chain config that we have for forc-node to the disk as the fuel-core cli expects a path to the config.


re: Network Updates:

In my understanding currently users needs to go and fetch the updated chain config from github, and point to that while running the node if the network config changes. So this is already happening(?) I feel like this automates that workflow. 

Currently if there is an update to the network config, they need to fetch the updated network config from github manually. With this they just need to update forc-node version, which will start using the updated network config files as I was planning to cut a new release of forc-node as we change the network params.

Other option is to check online to see if the config is changed, this enables updates without needing to update the forc-node. I felt like explicit changes with the versioning would be more cleaner but if added ux improvement has more upside than the downside of shadowing the actual changes happening in the configurations we can do that.

Reply:
My first thought was that we needed to have chain config by default with forc node because If we do not provide them out of the box, then running testnet/mainnet nodes through forc-node will still be involved. (They need to first fetch the config)

Reply:
I guess I'm a bit unclear on this code - does it create a local copy that the user is expected to modify? Or is it just writing files to disk and ensuring they're up-to-date?

We definitely need to provide chain config files to users, otherwise this tool won't be much use. But for the networks that we run (testnet & mainnet), users will get no where modifying the chain config files, and IMO it is not an option we should even give them. If users want to start modifying things, that's the point where they can just run a node directly and figure it out themselves.

I just want to make sure we're not creating copies of these files on disk with no plans to ensure they're up-to-date. We need to make sure for testnet & mainnet that the most recent chain config is used on launch, especially if an update has happened since the last time the user ran the tool.

Reply:
Oh no this is not for the users to modify actually, this is for unpacking what we embed in to the binary as a folder to the disk so that fuel-core can read from. 

It writes to the disk so that we are sure, the chain config read by the fuel-core is the one provided via the forc-node. Once we update the embedded config for forc-node and cut a new release, the new forc-node version will be making sure that the new config will be used.

So current flow to update config is:

1. Update network config in sway repo
2. Cut a new release
3. Point the (TESTNET/MAINNET) channel to the new release of forc-node

if users are using the latest version of the fuelup channel they will have the up-to date network config.

An alternative to this is to check online the central network config repo to see if there is an update and if so fetching it. This removes the need for cutting a new release of forc-node each time we update the network config. But hides the fact that network config is changing.

We can the the 2nd option too if that sounds preferable.

Reply:
I implemented a online checker that checks for updates in testnet and mainnet.  For local config, if it doesn't exists we fetch one, but after that we do not check for updates and let the user change it as they see fit and we do not override it in any way.